### ISSUE TITLE
Error when `parallel = TRUE`

### ISSUE THREAD

**Issue by JeffreyRStevens** [FBartos/RoBMA#25]
I saw in the documentation that the parallel implementation is not completely stable for the `RoBMA()` function. But it would be nice to use, if possible! When I try to use `parallel = TRUE`, I receive the following error:
``
Error in parallel::makePSOCKcluster(floor(RoBMA.get_option("max_cores")/object$fit_control[["chains"]])) : 
  numeric 'names' must be >= 1
``
This occurs without setting up any clusters but also when I set `options(mc.cores = parallel::detectCores())` or `parallel::makeCluster(parallel::detectCores())`. Am I doing something wrong? Or is there progress on making parallel implementation stable?
I'm using R v4.2.2, RoBMA v2.3.1, Ubuntu v22.04 (with 24 cores).
Thanks!

**Comment by FBartos** [comment]
Hi Jeffrey,

Thanks for reporting the bug. To be honest, I never came across anything like it before.

In RoBMA, I'm using the `parallel` package to handle the multithreading. The fact that even `parallel::makeCluster(parallel::detectCores())` crashes the server seems to be due to some interaction of the `parallel` package and your server setting. I wanted to send you their forum for help, but I couldn't quickly find any. I would probably try mailing the R (https://stat.ethz.ch/mailman/listinfo/r-help)?

Cheers,
Frantisek

**Comment by JeffreyRStevens** [comment]
Thanks, Frantisek. I'll look into that. Meanwhile, I just came across this recent article on [Please Avoid detectCores() in your R Packages](https://www.jottr.org/2022/12/05/avoid-detectcores/) that recommends replacing all `parallel::detectCores()` with `parallelly::availableCores()` or define `ncores <- max(1L, parallel::detectCores(), na.rm = TRUE)`, which could be adapted to your `max_cores`. Hopefully, tomorrow I can clone your package and try this out.

**Comment by FBartos** [comment]
Thanks for the suggestion! I will check it out.

My first thought is that using:
```max(1L, parallel::detectCores(), na.rm = TRUE)```
will ascertain the program does not crash if it's unable to detect the cores, however, the it will run single-core only, which is not the desired behavior either.

Could you maybe try calling `RoBMA.get_option("max_cores")`? I guess it will be NA.
Then setting `RoBMA.options(max_cores = 23)` and running RoBMA with `parallel = TRUE` again?

If that helps, and you also get proper number of cores from `parallelly::availableCores()`, I will switch the implementation.

Thanks for the additional details!
Frantisek

**Comment by JeffreyRStevens** [comment]
Many thanks for the response.

You're correct--`max(1L, parallel::detectCores(), na.rm = TRUE)` will likely just use a single core, so not a great solution. Maybe `parallelly::availableCores()` fixes this, I'm not sure.

When I run `RoBMA.get_option("max_cores")`, I get 1. When I run `parallel::detectCores()`, I get 24. So it seems like something in the assignment of `max_cores` is causing the problem.

When I set `RoBMA.options(max_cores = 23)` and `parallel = TRUE`, it works! So that is definitely the workaround for me for now--thank you!

In poking around a bit, I noticed a few things. If I install RoBMA from CRAN  with `install.packages("RoBMA")` or from GitHub with `devtools::install_github("fbartos/RoBMA")` is when I get this error (`RoBMA.get_option("max_cores")` is 1 in both cases). When I cloned the repo locally, built the source package, and load the local version, the parallel version runs fine without needing to define `max_cores` and `RoBMA.get_option("max_cores")` is 23. So I can't test whether `parallelly::availableCores()` works better because the local build works fine.

In summary, I can circumvent the problem by assigning `RoBMA.options(max_cores = 23)`. But there seems to be an issue with the package builds on CRAN and GitHub maybe?

**Comment by FBartos** [comment]
This is very odd, and I have no idea what can be wrong. When I install the package from CRAN and check `RoBMA.get_option("max_cores")`, I get the very wrong number of 55 (I wish I had such a CPU lol). When I run `parallel::detectCores()`, I get the correct number though (and the same when I compile the package locally).

This is definitely not ideal as the package thinks it has more cores than it actually does and does not distribute the treads in the most efficient way. I will check what else can be done with this.

**Comment by smith558** [comment]
Also affected by this issue.
